File: objectModel/Python/cdm/objectmodel/cdm_document_collection.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from typing import Optional, List, Union, TYPE_CHECKING

from cdm.enums import CdmObjectType, CdmLogCode
from cdm.utilities import logger

from .cdm_collection import CdmCollection

if TYPE_CHECKING:
    from .cdm_document_def import CdmDocumentDefinition

_TAG = 'CdmDocumentCollection'

class CdmDocumentCollection(CdmCollection):
    def __init__(self, ctx: 'CdmCorpusContext', owner: 'CdmObject'):
        super().__init__(ctx, owner, CdmObjectType.DOCUMENT_DEF)

    def append(self, document: Union[str, 'CdmDocumentDefinition'], document_name: Optional[str] = None) -> 'CdmDocumentDefinition':
        if isinstance(document, str):
            return self.append(self.ctx.corpus.make_object(self.default_type, document))

        if document_name:
            document.name = document_name

        if not self._check_and_add_item_modifications(document):
            return None
        
        document.owner = self.owner
        list.append(self, document)

        return document

    def clear(self) -> None:
        for element in self:
            self._remove_item_modifications(element.name)
        super().clear()

    def insert(self, index: int, document: 'CdmDocumentDefinition') -> None:
        if not self._check_and_add_item_modifications(document):
            return
        
        document.owner = self.owner
        list.insert(self, index, document)

    def remove(self, doc: Union[str, 'CdmDocumentDefinition']) -> None:
        """Removes a document from this folder object.

        arguments:
        name: The name of the document."""
        name = doc if isinstance(doc, str) else doc.name
        if name in self.owner._document_lookup:
            self._remove_item_modifications(name)
            # setting this currentlyResolving flag will keep the base collection code from setting the inDocument to null
            # this makes sense because a document is "in" itself. always.
            b_save = self.ctx.corpus._is_currently_resolving
            self.ctx.corpus._is_currently_resolving = True
            elem = next(d for d in self if d.name == name)
            self.ctx.corpus._is_currently_resolving = b_save
            return super().remove(elem)

    def _check_and_add_item_modifications(self, document: 'CdmDocumentDefinition') -> bool:
        if self.item(document.name) is not None:
            logger.error(self.ctx, _TAG, '_check_and_add_item_modifications', document.at_corpus_path,
                       CdmLogCode.ERR_DOC_ALREADY_EXIST, document.name,
                         lambda x:self.owner.at_corpus_path if self.owner.at_corpus_path is not None else self.owner.name)
            return False;
        
        if document.owner and document.owner is not self.owner:
            # this is fun! the document is moving from one folder to another
            # it must be removed from the old folder for sure, but also now
            # there will be a problem with any corpus paths that are relative to that old folder location.
            # so, whip through the document and change any corpus paths to be relative to this folder
            document._localize_corpus_paths(self.owner)  # returns false if it fails, but ... who cares? we tried
            document.owner.documents.remove(document.name)

        document._folder_path = self.owner._folder_path
        document.owner = self.owner
        document._namespace = self.owner._namespace

        super()._make_document_dirty()  # set the document to dirty so it will get saved in the new folder location if saved
        self.owner._corpus._add_document_objects(self.owner, document)
        return True

    def _remove_item_modifications(self, name: str) -> None:
        self.owner._corpus._remove_document_objects(self.owner, self.owner._document_lookup[name])
        self.owner._document_lookup.pop(name)

File: objectModel/TypeScript/Cdm/CdmDocumentCollection.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { isBoolean, isString } from 'util';
import {
    CdmCollection,
    CdmCorpusContext,
    CdmDocumentDefinition,
    CdmFolderDefinition,
    cdmObjectType,
    Logger,
    cdmLogCode
} from '../internal';

export class CdmDocumentCollection extends CdmCollection<CdmDocumentDefinition> {
    private TAG: string = CdmDocumentCollection.name;

    constructor(ctx: CdmCorpusContext, owner: CdmFolderDefinition) {
        super(ctx, owner, cdmObjectType.documentDef);
    }

    protected get owner(): CdmFolderDefinition {
        return super.owner as CdmFolderDefinition;
    }

    protected set owner(value: CdmFolderDefinition) {
        super.owner = value;
    }

    /**
     * @inheritdoc
     */
    public push(documentOrName: string | CdmDocumentDefinition, documentNameOrSimpleRef?: string | boolean)
        : CdmDocumentDefinition {
        let document: CdmDocumentDefinition;
        if (documentOrName !== undefined && isString(documentOrName)) {
            const simpleRef: boolean = isBoolean(documentNameOrSimpleRef) ? documentNameOrSimpleRef : false;
            document = this.ctx.corpus.MakeObject<CdmDocumentDefinition>(this.defaultType, documentOrName, simpleRef);
        } else if (documentOrName instanceof CdmDocumentDefinition) {
            document = documentOrName;
        }

        if (document) {
            if (documentNameOrSimpleRef !== undefined && isString(documentNameOrSimpleRef)) {
                document.name = documentNameOrSimpleRef;
            }

            if (!this.checkAndAddItemModifications(document)) {
                return undefined;
            }

            // why is this collection unlike all other collections?
            // because documents are in folders. folders are not in documents.
            document.owner = this.owner;
            this.allItems.push(document);

            return document;
        }
    }

    public insert(index: number, docDef: CdmDocumentDefinition): void {
        if (!this.checkAndAddItemModifications(docDef)) {
            return;
        }

        // why is this collection unlike all other collections?
        // because documents are in folders. folders are not in documents.
        docDef.owner = this.owner;
        const l: number = this.length;
        for (let i: number = l; i > index; i = i - 1) {
            this.allItems[i] = this.allItems[i - 1];
        }
        this.allItems[index] = docDef;
    }

    public remove(object: string | CdmDocumentDefinition): boolean {
        if (isString(object)) {
            if (this.owner.documentLookup.has(object)) {
                this.removeItemModifications(object);
                const index: number = this.allItems.findIndex((elem: CdmDocumentDefinition) => elem.name === object);
                // setting this currentlyResolving flag will keep the base collection code from setting the inDocument to null
                // this makes sense because a document is "in" itself. always.
                const bSave: boolean = this.ctx.corpus.isCurrentlyResolving;
                this.ctx.corpus.isCurrentlyResolving = true;
                super.removeAt(index);
                this.ctx.corpus.isCurrentlyResolving = bSave;

                return true;
            }

            return false;
        } else {
            return this.remove(object.name);
        }
    }

    public removeAt(index: number): void {
        if (index >= 0 && index < this.length) {
            this.remove(this.allItems[index].name);
        }
    }

    public clear(): void {
        this.allItems.forEach((document: CdmDocumentDefinition) => {
            this.removeItemModifications(document.name);
        });
        super.clear();
    }

    private checkAndAddItemModifications(document: CdmDocumentDefinition): boolean {
        if (this.item(document.name) !== undefined) {
            Logger.error(this.ctx, this.TAG, "checkAndAddItemModifications", document.atCorpusPath, cdmLogCode.ErrDocAlreadyExist, document.name,
            this.owner.atCorpusPath != undefined ? this.owner.atCorpusPath : this.owner.name);
            return false;
        }

        if (document.owner !== undefined && document.owner !== this.owner) {
            // this is fun! the document is moving from one folder to another
            // it must be removed from the old folder for sure, but also now
            // there will be a problem with any corpus paths that are relative to that old folder location.
            // so, whip through the document and change any corpus paths to be relative to this folder
            document.localizeCorpusPaths(this.owner); // returns false if it fails, but ... who cares? we tried
            (document.owner as CdmFolderDefinition).documents.remove(document.name);
        }
        document.folderPath = this.owner.folderPath;
        document.owner = this.owner;
        document.namespace = this.owner.namespace;
        this.makeDocumentDirty(); // set the document to dirty so it will get saved in the new folder location if saved
        this.owner.corpus.addDocumentObjects(this.owner, document);

        return true;
    }

    private removeItemModifications(documentName: string): void {
        this.owner.corpus.removeDocumentObjects(this.owner, this.owner.documentLookup.get(documentName));
        this.owner.documentLookup.delete(documentName);
    }
}
